Add ITuple checks to PSGetIndexBinder

SeeminglyScience opened the pull request:
## PR Summary

Resolves #7632.

This change enables index operations on objects that implement `ITuple`, including slicing and negative indexing.

Add checks in `PSGetIndexBinder` for types that implement an interface with `DefaultMemberAttribute` declared, such as `ITuple`. `Tuple` and `ValueTuple` implement `ITuple` interface, and thus they are made indexable by this PR.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

SeeminglyScience committed:
Add ITuple checks to PSGetIndexBinder

This change enables index operations on objects that implement ITuple.

lzybkr commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -3946,6 +3946,11 @@ public override DynamicMetaObject FallbackGetIndex(DynamicMetaObject target, Dyn
                 return InvokeIndexer(target, indexes, errorSuggestion, defaultMember.MemberName).WriteToDebugLog(this);
             }
 
+            if (typeof(ITuple).IsAssignableFrom(target.LimitType))
```
This doesn't feel like the right fix. `ITuple` specifies a `DefaultMemberAttribute`, but we don't find it above because `GetCustomAttributes` is only checking classes, not implemented interfaces.

The correct fix is to search interfaces for `DefaultMemberAttribute` - there are other interfaces that specify `DefaultMemberAttribute`, and some of those are not currently handled correctly:

```
System.Collections.IDictionary
System.Collections.IList
System.Collections.Generic.IDictionary`2
System.Collections.Generic.IList`1
System.Collections.Generic.IReadOnlyDictionary`2
System.Collections.Generic.IReadOnlyList`1
System.Runtime.CompilerServices.ITuple
System.Runtime.CompilerServices.IRuntimeVariables
System.Collections.Specialized.IOrderedDictionary
System.Linq.ILookup`2
System.Data.IColumnMappingCollection
System.Data.IDataParameterCollection
System.Data.IDataRecord
System.Data.ITableMappingCollection
System.ComponentModel.IDataErrorInfo
Newtonsoft.Json.Linq.IJEnumerable`1
```

SeeminglyScience replied:
@lzybkr Good call! I had checked that none of the default `ITuple` implementations had that attribute but didn't think to check the interface itself.

I've updated the change to look for `DefaultMemberAttribute` in the loop that looks for `IDictionary<,>`.  I tested the update with a custom `ILookup<,>` with an explicitly implemented indexer, that didn't work previously but does now.

iSazonov replied:
Do now we support all these interfaces? If so, it's good to have a comment in the code and tests (or even extend the tests)
Also we need to document this.

SeeminglyScience replied:
@iSazonov They were already supported.  What *wasn't* supported was indexing into classes that *explicitly* implement the indexer of an interface and define no other indexer.

All `ITuple` implementations explicitly implement the `ITuple` indexer, so while indexing tuples was technically supported, it wasn't working.

iSazonov replied:
@SeeminglyScience Sorry, I did not ask the question correctly. Your explanation is useful and I just believe that it will make it easier to understand the binder if it's in the code.

SeeminglyScience replied:
@iSazonov Ah yes absolutely, I'll add that tonight.

SeeminglyScience replied:
@iSazonov I've added a comment explaining purpose of the extra check in the loop.

SeeminglyScience committed:
Check for explicitly implemented indexers

Remove the logic to specifically check for ITuple implementations.
Instead, search all implemented interfaces for a
DefaultMemberAttribute. This will ensure all explicitly implemented
indexers are found.

SeeminglyScience committed:
Fix Count/Length property search order

lzybkr commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -4046,7 +4056,12 @@ internal static bool CanIndexFromEndWithNegativeIndex(DynamicMetaObject target)
+                return true;
+            }
+
+            return typeof(ITuple).IsAssignableFrom(limitType);
```
I wonder if this whole method shouldn't just check for a `Count` or `Length` property.

SeeminglyScience replied:
I think that would be a breaking change.  For example, if a `Dictionary<,>` could make it that far into the binder (I know it can't, but custom code might) then `-1` could be a valid key.

I did a quick scan of the types visible by default in RC1, looking for types that:

1. Are decorated with `DefaultMemberAttribute`
1. Do not implement `ICollection` or `ICollection<>` (as the `Count` property coming from these make it likely enough to work as expected)
1. Do not implement `IDictionary<,>` (as they'll be handled earlier in the binder)

There's a couple types that jump out at a glance as *possibly* an issue, depending on how imperative it is to preserve that behavior. 

Here's the list in full, a couple obvious ones that jump out are `System.Linq.ILookup<,>`, `ReadOnlyPSMemberInfoCollection<>`, and some CIM related classes like `CimMethodParameterCollection`. 

```
System.String
System.ReadOnlySpan`1[T]
System.Span`1[T]
System.Numerics.Vector`1[T]
System.Text.StringBuilder
System.Runtime.CompilerServices.ITuple
System.Management.Automation.PSEventArgsCollection
System.Management.Automation.ReadOnlyPSMemberInfoCollection`1[T]
System.Management.Automation.Runspaces.InitialSessionStateEntryCollection`1[T]
System.Runtime.CompilerServices.IRuntimeVariables
Microsoft.Management.Infrastructure.CimMethodParametersCollection
Microsoft.Management.Infrastructure.Generic.CimKeyedCollection`1[T]
Microsoft.Management.Infrastructure.Generic.CimReadOnlyKeyedCollection`1[T]
System.Xml.XmlCDataSection
System.Xml.XmlCharacterData
System.Xml.XmlComment
System.Xml.XmlNodeList
System.Xml.XmlSignificantWhitespace
System.Xml.XmlText
System.Xml.XmlWhitespace
System.Xml.Serialization.XmlMembersMapping
System.Xml.Schema.XmlSchemaObjectTable
System.Collections.Specialized.StringDictionary
System.Linq.ILookup`2[TKey,TElement]
System.Linq.Lookup`2[TKey,TElement]
System.Data.SqlTypes.SqlChars
System.Data.SqlTypes.SqlBinary
System.Data.SqlTypes.SqlBytes
System.ComponentModel.MaskedTextProvider
Microsoft.CodeAnalysis.ChildSyntaxList
Microsoft.CodeAnalysis.SeparatedSyntaxList`1[TNode]
Microsoft.CodeAnalysis.SyntaxList`1[TNode]
Microsoft.CodeAnalysis.SyntaxNodeOrTokenList
Microsoft.CodeAnalysis.SyntaxTokenList
Microsoft.CodeAnalysis.SyntaxTriviaList
Microsoft.CodeAnalysis.Text.SourceText
Microsoft.CodeAnalysis.Text.TextLineCollection
System.Reflection.Metadata.GenericParameterHandleCollection
System.Reflection.Metadata.GenericParameterConstraintHandleCollection
Markdig.Helpers.StringSlice
```

While not incredibly likely to actually break existing code, perhaps it should be something the committee discusses?

lzybkr replied:
The risk is that an indexer supports negative values as an index somehow, e.g. as the key in a dictionary.

Based on your list, `StringDictionary`, ``Lookup`2`` and ``ILookup`2`` do look like they'd support a negative value as the index.

Maybe the check should be - the indexer accepts a single `int` and is not a generic parameter. This avoids the problem with `StringDictionary` with the signature `string Item(string key) {get;set;}` and `ILookup[int,T]]` where even though the indexer takes an `int`, it was generic and therefore not always an int.

SeeminglyScience committed:
Generalized can index with negative check

Changed the method CanIndexFromEndWithNegativeIndex to determine
if negative indexing should be allowed to be more generalized.
Previously this was implemented by checking to see if the limit
type matched a specific set of hard coded types. This was updated
to allow negative indexing if the following is true.

- The best matching index method takes only one parameter and it is
  typed as an int

- The index parameter is not a generic type parameter of a constructed
  generic type

- A "Count" or "Length" property it also found

SeeminglyScience committed:
Added comment for additional default member check

SeeminglyScience committed:
Merge branch 'master' into enable-tuple-index-ops

SeeminglyScience committed:
Fix is generic parameter logic for negative index

iSazonov commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -4027,26 +4040,21 @@ private DynamicMetaObject CannotIndexTarget(DynamicMetaObject target, DynamicMet
-            }
-
-            if (typeof(OrderedDictionary).IsAssignableFrom(limitType))
+            if (getterParams.Length != 1 || getterParams[0].ParameterType != typeof(int))
```
Can getterParams be null?

lzybkr replied:
No, reflection will return empty arrays but not null.

iSazonov commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -4242,6 +4249,7 @@ private DynamicMetaObject GetIndexMultiDimensionArray(DynamicMetaObject target,
                 }
             }
 
+
```
Please remove the extra line.

iSazonov commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -4258,13 +4266,16 @@ private DynamicMetaObject GetIndexMultiDimensionArray(DynamicMetaObject target,
+                if (lengthProperty == null)
+                {
+                    lengthProperty = target.LimitType.GetProperty("Count") ??
+                                     target.LimitType.GetProperty("Length"); // for string
```
Comment should be on separate lines. Please remove the comment or add enhanced comment above.

SeeminglyScience replied:
Removed the inline comment and added a more description comment above.

lzybkr commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -4258,7 +4248,7 @@ private DynamicMetaObject GetIndexMultiDimensionArray(DynamicMetaObject target,
-            if (getterParams.Length == 1 && getterParams[0].ParameterType == typeof(int) && CanIndexFromEndWithNegativeIndex(target))
+            if (CanIndexFromEndWithNegativeIndex(target, getter, getterParams))
             {
                 // PowerShell supports negative indexing for some types (specifically, types implementing IList or IList<T>).
```
This comment should be updated.

SeeminglyScience replied:
Good catch, I've updated the comment to fully explain the supported scenarios.

lzybkr reviewed (approved):
LGTM, but it'd be good to have @daxian-dbw sign off before merging.

SeeminglyScience committed:
Add additional comments and fix extra new line

iSazonov commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -4268,13 +4267,25 @@ private DynamicMetaObject GetIndexMultiDimensionArray(DynamicMetaObject target,
             if (CanIndexFromEndWithNegativeIndex(target, getter, getterParams))
             {
-                // PowerShell supports negative indexing for some types (specifically, types implementing IList or IList<T>).
+                // PowerShell supports negative indexing for for types that meet the following criteria:
```
Typo - for for.

SeeminglyScience committed:
Fix typo and change comment formatting

iSazonov commented:
LGTM.

daxian-dbw commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -4027,26 +4040,21 @@ private DynamicMetaObject CannotIndexTarget(DynamicMetaObject target, DynamicMet
 
-        internal static bool CanIndexFromEndWithNegativeIndex(DynamicMetaObject target)
+        internal static bool CanIndexFromEndWithNegativeIndex(
+            DynamicMetaObject target,
```
Turns out this parameter is not used within the method.

I think we probably should keep the existing checks for the common scenarios, since `Module.ResolveMethod` is relatively expensive.

daxian-dbw commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -4258,13 +4265,23 @@ private DynamicMetaObject GetIndexMultiDimensionArray(DynamicMetaObject target,
+                //        (this is to exclude indexers for types that could use a negative index as
+                //        a valid key like System.Linq.ILookup)
+                //      - Declares a "Count" or "Length" property
+                //      - Does not inherit from IDictionary<> as that is handled earlier in the binder
```
Great comments! I think they should be moved to the method `CanIndexFromEndWithNegativeIndex`  😄

daxian-dbw commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -3928,6 +3928,8 @@ public override DynamicMetaObject FallbackGetIndex(DynamicMetaObject target, Dyn
                 return GetIndexArray(target, indexes, errorSuggestion).WriteToDebugLog(this);
             }
 
+            var defaultMember = target.LimitType.GetCustomAttributes<DefaultMemberAttribute>(true).FirstOrDefault();
```
Can we change this to also use `GetCustomAttribute<DefaultMemberAttribute>(inherit: true)`?

SeeminglyScience replied:
Hmm actually it looks like `DefaultMemberAttribute` can be declared multiple times, and this method will throw in that case.  I believe I need to change the other instance to match this one instead.

daxian-dbw replied:
Oh, that sounds good. Thanks for clarifying.

daxian-dbw commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -3938,12 +3940,23 @@ public override DynamicMetaObject FallbackGetIndex(DynamicMetaObject target, Dyn
+                // This check will catch those cases.
+                if (defaultMember == null)
+                {
+                    defaultMember = i.GetCustomAttribute<DefaultMemberAttribute>(inherit: true);
```
A minor comment: Attributes cannot be inherited from interfaces, so it should be `inherit: false`.

SeeminglyScience commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -4027,26 +4040,21 @@ private DynamicMetaObject CannotIndexTarget(DynamicMetaObject target, DynamicMet
+            // parameter is a generic type parameter. Module.ResolveMethod is used
+            // because the indexer could be a method from a constructed generic type.
+            MethodBase baseMethod = indexer.Module.ResolveMethod(indexer.MetadataToken);
+            return !baseMethod.GetParameters()[0].ParameterType.IsGenericParameter;
```
That tests return type as well.  The indexer for something like `IList<>` would also be excluded.

daxian-dbw replied:
Good point. `ContainsGenericParameters` indicates if this method comes from a generic type.

SeeminglyScience committed:
Add original type checks back to CanNegativeIndex

SeeminglyScience committed:
Change method to get attributes and move comment

daxian-dbw committed:
[Feature] Rearrange the check order in CanIndexFromEndWithNegativeIndex

daxian-dbw commented:
@SeeminglyScience I just pushed a commit to re-arrange the order of the checks in `CanIndexFromEndWithNegativeIndex` (parameter check is simpler), and also add `[Feature]` tag to run all the tests. Hope you don't mind!

daxian-dbw reviewed (approved):
LGTM

iSazonov commented:
@SeeminglyScience Thanks for the great enhancement! 
 Please update the PR description to facilitate documentation.

daxian-dbw commented:
I have updated the PR description. Will merge the PR.
